File: src/utils/matlab/libs/plotKalman.m
function plotKalman(channelList, trackResults, settings)
% This function plots the tracking results for the given channel list.
%
% plotTracking(channelList, trackResults, settings)
%
%   Inputs:
%       channelList     - list of channels to be plotted.
%       trackResults    - tracking results from the tracking function.
%       settings        - receiver settings.

%--------------------------------------------------------------------------
%                           SoftGNSS v3.0
%
% Copyright (C) Darius Plausinaitis
% Written by Darius Plausinaitis
%--------------------------------------------------------------------------
%This program is free software; you can redistribute it and/or
%modify it under the terms of the GNU General Public License
%as published by the Free Software Foundation; either version 2
%of the License, or (at your option) any later version.
%
%This program is distributed in the hope that it will be useful,
%but WITHOUT ANY WARRANTY; without even the implied warranty of
%MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%GNU General Public License for more details.
%
%You should have received a copy of the GNU General Public License
%along with this program; if not, write to the Free Software
%Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
%USA.
%--------------------------------------------------------------------------

% Protection - if the list contains incorrect channel numbers
channelList = intersect(channelList, 1:settings.numberOfChannels);

%=== For all listed channels ==============================================
for channelNr = channelList
    
    %% Select (or create) and clear the figure ================================
    % The number 200 is added just for more convenient handling of the open
    % figure windows, when many figures are closed and reopened.
    % Figures drawn or opened by the user, will not be "overwritten" by
    % this function.
    
    figure(channelNr +200);
    clf(channelNr +200);
    set(channelNr +200, 'Name', ['Channel ', num2str(channelNr), ...
        ' (PRN ', ...
        num2str(trackResults(channelNr).PRN(end-1)), ...
        ') results']);
    
    timeStart = settings.timeStartInSeconds;
    
    %% Draw axes ==============================================================
    % Row 1
    handles(1, 1) = subplot(4, 2, 1);
    handles(1, 2) = subplot(4, 2, 2);
    % Row 2
    handles(2, 1) = subplot(4, 2, 3);
    handles(2, 2) = subplot(4, 2, 4);
    % Row 3
    handles(3, 1) = subplot(4, 2, [5 6]);
    % Row 4
    handles(4, 1) = subplot(4, 2, [7 8]);
    
    %% Plot all figures =======================================================
    
    timeAxisInSeconds = (1:settings.msToProcess)/1000;
    
    %----- CNo for signal----------------------------------
    plot  (handles(1, 1), timeAxisInSeconds, ...
        trackResults(channelNr).CNo(1:settings.msToProcess), 'b');
    
    grid  (handles(1, 1));
    axis  (handles(1, 1), 'tight');
    xlabel(handles(1, 1), 'Time (s)');
    ylabel(handles(1, 1), 'CNo (dB-Hz)');
    title (handles(1, 1), 'Carrier to Noise Ratio');
    
    %----- PLL discriminator filtered----------------------------------
    plot  (handles(1, 2), timeAxisInSeconds, ...
        trackResults(channelNr).state1(1:settings.msToProcess), 'b');
    
    grid  (handles(1, 2));
    axis  (handles(1, 2), 'tight');
    xlim  (handles(1, 2), [timeStart, timeAxisInSeconds(end)]);
    xlabel(handles(1, 2), 'Time (s)');
    ylabel(handles(1, 2), 'Phase Amplitude');
    title (handles(1, 2), 'Filtered Carrier Phase');
    
    %----- Carrier Frequency --------------------------------
    plot  (handles(2, 1), timeAxisInSeconds(2:end), ...
        trackResults(channelNr).state2(2:settings.msToProcess), 'Color',[0.42 0.25 0.39]);
    
    grid  (handles(2, 1));
    axis  (handles(2, 1));
    xlim  (handles(2, 1), [timeStart, timeAxisInSeconds(end)]);
    xlabel(handles(2, 1), 'Time (s)');
    ylabel(handles(2, 1), 'Freq (hz)');
    title (handles(2, 1), 'Filtered Doppler Frequency');
    
    %----- Carrier Frequency Rate --------------------------------
    plot  (handles(2, 2), timeAxisInSeconds(2:end), ...
        trackResults(channelNr).state3(2:settings.msToProcess), 'Color',[0.42 0.25 0.39]);
    
    grid  (handles(2, 2));
    axis  (handles(2, 2));
    xlim  (handles(2, 2), [timeStart, timeAxisInSeconds(end)]);
    xlabel(handles(2, 2), 'Time (s)');
    ylabel(handles(2, 2), 'Freq (hz)');
    title (handles(2, 2), 'Filtered Doppler Frequency Rate');
    
    %----- PLL discriminator unfiltered--------------------------------
    plot  (handles(3, 1), timeAxisInSeconds, ...
        trackResults(channelNr).innovation, 'r');
    
    grid  (handles(3, 1));
    axis  (handles(3, 1), 'auto');
    xlim  (handles(3, 1), [timeStart, timeAxisInSeconds(end)]);
    xlabel(handles(3, 1), 'Time (s)');
    ylabel(handles(3, 1), 'Amplitude');
    title (handles(3, 1), 'Raw PLL discriminator (Innovation)');
    
    
    %----- PLL discriminator covariance --------------------------------
    plot  (handles(4, 1), timeAxisInSeconds, ...
        trackResults(channelNr).r_noise_cov, 'r');
    
    grid  (handles(4, 1));
    axis  (handles(4, 1), 'auto');
    xlim  (handles(4, 1), [timeStart, timeAxisInSeconds(end)]);
    xlabel(handles(4, 1), 'Time (s)');
    ylabel(handles(4, 1), 'Variance');
    title (handles(4, 1), 'Estimated Noise Variance');
end % for channelNr = channelList

File: src/utils/python/lib/plotKalman.py
"""
 plotKalman.py
   plotKalman (channelNr, trackResults, settings)

 This function plots the tracking results for the given channel list.

 Irene Pérez Riega, 2023. iperrie@inta.es

      Args:
        channelList     - list of channels to be plotted.
        trackResults    - tracking results from the tracking function.
        settings        - receiver settings.

      Modifiable in the file:
        fig_path        - Path where doppler plots will be save

 -----------------------------------------------------------------------------

 GNSS-SDR is a Global Navigation Satellite System software-defined receiver.
 This file is part of GNSS-SDR.

 Copyright (C) 2022  (see AUTHORS file for a list of contributors)
 SPDX-License-Identifier: GPL-3.0-or-later

 -----------------------------------------------------------------------------
"""

import matplotlib.pyplot as plt
import numpy as np
import os


def plotKalman(channelNr, trackResults, settings):

    # ---------- CHANGE HERE:
    fig_path = '/home/labnav/Desktop/TEST_IRENE/PLOTS/PlotKalman'

    if not os.path.exists(fig_path):
        os.makedirs(fig_path)

    # Protection - if the list contains incorrect channel numbers
    channelNr = np.intersect1d(channelNr,
                               np.arange(1, settings['numberOfChannels'] + 1))

    for channelNr in channelNr:
        time_start = settings['timeStartInSeconds']
        time_axis_in_seconds = np.arange(1, settings['msToProcess']+1)/1000

        # Plot all figures
        plt.figure(figsize=(1920 / 100, 1080 / 100))
        plt.clf()
        plt.gcf().canvas.set_window_title(
            f'Channel {channelNr} (PRN '
            f'{str(trackResults[channelNr-1]["PRN"][-2])}) results')
        plt.subplots_adjust(left=0.1, right=0.9, top=0.9, bottom=0.1,
                            hspace=0.4, wspace=0.4)
        plt.tight_layout()

        # Row 1
        # ----- CNo for signal -----------------------------------------------
        # Measure of the ratio between carrier signal power and noise power
        plt.subplot(4, 2, 1)
        plt.plot(time_axis_in_seconds,
                 trackResults[channelNr-1]['CNo'][:settings['msToProcess']],
                 'b')
        plt.grid()
        plt.axis('tight')
        plt.xlabel('Time (s)')
        plt.ylabel('CNo (dB-Hz)')
        plt.title('Carrier to Noise Ratio', fontweight='bold')

        # ----- PLL discriminator filtered -----------------------------------
        plt.subplot(4, 2, 2)
        plt.plot(time_axis_in_seconds,
                 trackResults[channelNr-1]['state1']
                 [:settings['msToProcess']], 'b')
        plt.grid()
        plt.axis('tight')
        plt.xlim([time_start, time_axis_in_seconds[-1]])
        plt.xlabel('Time (s)')
        plt.ylabel('Phase Amplitude')
        plt.title('Filtered Carrier Phase', fontweight='bold')

        # Row 2
        # ----- Carrier Frequency --------------------------------------------
        # Filtered carrier frequency of (transmitted by a satellite)
        # for a specific channel
        plt.subplot(4, 2, 3)
        plt.plot(time_axis_in_seconds[1:],
                 trackResults[channelNr-1]['state2']
                 [1:settings['msToProcess']], color=[0.42, 0.25, 0.39])
        plt.grid()
        plt.axis('auto')
        plt.xlim(time_start, time_axis_in_seconds[-1])
        plt.xlabel('Time (s)')
        plt.ylabel('Freq (Hz)')
        plt.title('Filtered Carrier Frequency', fontweight='bold')

        # ----- Carrier Frequency Rate ---------------------------------------
        plt.subplot(4, 2, 4)
        plt.plot(time_axis_in_seconds[1:],
                 trackResults[channelNr-1]['state3']
                 [1:settings['msToProcess']], color=[0.42, 0.25, 0.39])
        plt.grid()
        plt.axis('auto')
        plt.xlim(time_start, time_axis_in_seconds[-1])
        plt.xlabel('Time (s)')
        plt.ylabel('Freq (Hz)')
        plt.title('Filtered Carrier Frequency Rate', fontweight='bold')

        # Row 3
        # ----- PLL discriminator unfiltered----------------------------------
        plt.subplot(4, 2, (5,6))
        plt.plot(time_axis_in_seconds,
                 trackResults[channelNr-1]['innovation'], 'r')
        plt.grid()
        plt.axis('auto')
        plt.xlim(time_start, time_axis_in_seconds[-1])
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.title('Raw PLL discriminator (Innovation)',fontweight='bold')

        # Row 4
        # ----- PLL discriminator covariance ---------------------------------
        plt.subplot(4, 2, (7,8))
        plt.plot(time_axis_in_seconds,
                 trackResults[channelNr-1]['r_noise_cov'], 'r')
        plt.grid()
        plt.axis('auto')
        plt.xlim(time_start, time_axis_in_seconds[-1])
        plt.xlabel('Time (s)')
        plt.ylabel('Variance')
        plt.title('Estimated Noise Variance', fontweight='bold')

        plt.tight_layout()
        plt.savefig(os.path.join(fig_path,
                                 f'kalman_ch{channelNr}_PRN_'
                                 f'{trackResults[channelNr - 1]["PRN"][-1]}'
                                 f'.png'))
        plt.show()
